pkg/login/social/generic_oauth_test.go

```diff
@@ -332,3 +332,97 @@ func TestUserInfoSearchesForEmailAndRole(t *testing.T) {
 		}
 	})
 }
+
+func TestUserInfoSearchesForLogin(t *testing.T) {
+	t.Run("Given a generic OAuth provider", func(t *testing.T) {
+		provider := SocialGenericOAuth{
+			SocialBase: &SocialBase{
+				log: log.New("generic_oauth_test"),
+			},
+			loginAttributePath: "login",
+		}
+
+		tests := []struct {
+			Name               string
+			APIURLResponse     interface{}
+			OAuth2Extra        interface{}
+			LoginAttributePath string
+			ExpectedLogin      string
+		}{
+			{
+				Name: "Given a valid id_token, a valid login path, no api response, use id_token",
+				OAuth2Extra: map[string]interface{}{
+					// { "login": "johndoe", "email": "john.doe@example.com" }
+					"id_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJsb2dpbiI6ImpvaG5kb2UiLCJlbWFpbCI6ImpvaG4uZG9lQGV4YW1wbGUuY29tIn0.sg4sRJCNpax_76XMgr277fdxhjjtNSWXKIOFv4_GJN8",
+				},
+				LoginAttributePath: "role",
+				ExpectedLogin:      "johndoe",
+			},
+			{
+				Name: "Given a valid id_token, no login path, no api response, use id_token",
+				OAuth2Extra: map[string]interface{}{
+					// { "login": "johndoe", "email": "john.doe@example.com" }
+					"id_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJsb2dpbiI6ImpvaG5kb2UiLCJlbWFpbCI6ImpvaG4uZG9lQGV4YW1wbGUuY29tIn0.sg4sRJCNpax_76XMgr277fdxhjjtNSWXKIOFv4_GJN8",
+				},
+				LoginAttributePath: "",
+				ExpectedLogin:      "johndoe",
+			},
+			{
+				Name: "Given no id_token, a valid login path, a valid api response, use api response",
+				APIURLResponse: map[string]interface{}{
+					"user_uid": "johndoe",
+					"email":    "john.doe@example.com",
+				},
+				LoginAttributePath: "user_uid",
+				ExpectedLogin:      "johndoe",
+			},
+			{
+				Name: "Given no id_token, no login path, a valid api response, use api response",
+				APIURLResponse: map[string]interface{}{
+					"login": "johndoe",
+				},
+				LoginAttributePath: "",
+				ExpectedLogin:      "johndoe",
+			},
+			{
+				Name: "Given no id_token, a login path, a valid api response without a login, use api response",
+				APIURLResponse: map[string]interface{}{
+					"username": "john.doe",
+				},
+				LoginAttributePath: "login",
+				ExpectedLogin:      "john.doe",
+			},
+			{
+				Name:               "Given no id_token, a valid login path, no api response, no data",
+				LoginAttributePath: "login",
+				ExpectedLogin:      "",
+			},
+		}
+
+		for _, test := range tests {
+			provider.loginAttributePath = test.LoginAttributePath
+			t.Run(test.Name, func(t *testing.T) {
+				response, err := json.Marshal(test.APIURLResponse)
+				require.NoError(t, err)
+				ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusOK)
+					w.Header().Set("Content-Type", "application/json")
+					_, err = io.WriteString(w, string(response))
```

Comment:
Isn't this more straightforward?
```suggestion
					_, err = io.Copy(w, response)
```

Reply:
Hm, in this case `response` should be `io.Reader`, not a `[]byte`.

Reply:
@alexanderzobnin Ah, I didn't notice that difference.

Reply:
In that case, wouldn't the following work?

```suggestion
_, err = w.Write(response)
```

Reply:
To be honest, I just copied it from the test above :). But I'll check it out.

Reply:
Aha, well unless I'm mistaken, going via `io.WriteString` is pointless.

Reply:
I made a [PR](https://github.com/grafana/grafana/pull/26679) to change the existing test to use `w.Write`.